File: app/src/main/java/com/droidrank/tictactoe/domain/BoardController.java
package com.droidrank.tictactoe.domain;

public class BoardController {

    private Board board;
    private String player1Symbol;
    private String player2Symbol;

    public BoardController(Board board, String player1Symbol, String player2Symbol) {
        this.board = board;
        this.player1Symbol = player1Symbol;
        this.player2Symbol = player2Symbol;
    }

    public Board getBoard() { return board; }

    public void reset() {
        board.reset();
    }

    public void selectBlock(int x, int y) {
        if (board.isPlayer1Turn() && board.isEmpty(x, y)) {
            board.setPosition(x, y, player1Symbol);
            board.setStatus(BoardStatus.PLAYER_2_MOVES);
        } else if (board.isPlayer2Turn() && board.isEmpty(x, y)) {
            board.setPosition(x, y, player2Symbol);
            board.setStatus(BoardStatus.PLAYER_1_MOVES);
        }
        checkFull();
        checkLine();
    }

    private void checkFull() { if (board.isFull()) board.setStatus(BoardStatus.DRAW); }

    private void checkLine() {
        String symbol;
        String[][] blocks = board.getBlocks();
        for (int row = 0; row < Board.ROWS; row++) {
            symbol = checkRow(blocks, row);
            if (symbol != null) {
                setWinner(symbol);
                return;
            }
        }
        for (int col = 0; col < Board.COLUMNS; col++) {
            symbol = checkColumn(blocks, col);
            if (symbol != null) {
                setWinner(symbol);
                return;
            }
        }
        symbol = checkDiagonals(blocks);
        if (symbol != null) setWinner(symbol);
    }

    private String checkColumn(String[][] blocks, int rowPos) {
        String[] row = blocks[rowPos];
        String symbol = row[0];
        for (int pos = 1; pos < row.length; pos++) {
            if (row[pos] == null || !row[pos].equals(symbol)) return null;
        }
        return symbol;
    }

    private String checkRow(String[][] blocks, int columnPos) {
        String symbol = blocks[0][columnPos];
        for (int pos = 1; pos < blocks.length; pos++) {
            if (blocks[pos][columnPos] == null || !blocks[pos][columnPos].equals(symbol)) return null;
        }
        return symbol;
    }

    private String checkDiagonals(String[][] blocks) {
        String symbol = blocks[0][0];
        for (int i = 1; i < blocks.length; i++) {
            if (blocks[i][i] == null || !blocks[i][i].equals(symbol)) symbol = null;
        }
        if (symbol != null) return symbol;
        symbol = blocks[0][blocks.length - 1];
        for (int i = 1; i < blocks.length; i++) {
            if (blocks[i][blocks.length - i - 1] == null || !blocks[i][blocks.length - i - 1].equals(symbol)) symbol = null;
        }
        if (symbol != null) return symbol;
        return null;
    }

    private void setWinner(String symbol) {
        if (player1Symbol.equals(symbol)) {
            board.setStatus(BoardStatus.PLAYER_1_WINS);
        } else if (player2Symbol.equals(symbol)) {
            board.setStatus(BoardStatus.PLAYER_2_WINS);
        }
    }
}

File: app/src/main/java/com/droidrank/tictactoe/domain/BoardController.kt
package com.droidrank.tictactoe.domain

class BoardController(
        val board: Board,
        private val player1Symbol: String,
        private val player2Symbol: String) {

    fun reset() = board.reset()

    fun selectBlock(x: Int, y: Int) {
        if (board.isPlayer1Turn && board.isEmpty(x, y)) {
            board.setPosition(x, y, player1Symbol)
            board.status = BoardStatus.PLAYER_2_MOVES
        } else if (board.isPlayer2Turn && board.isEmpty(x, y)) {
            board.setPosition(x, y, player2Symbol)
            board.status = BoardStatus.PLAYER_1_MOVES
        }
        checkFull()
        checkLine()
    }

    private fun checkFull() = with(board) {
        if (isFull) status = BoardStatus.DRAW
    }

    private fun checkLine() {
        var symbol: String?
        val blocks = board.blocks
        for (row in 0 until Board.ROWS) {
            symbol = checkRow(blocks, row)
            symbol?.let { setWinner(it) }
        }
        for (col in 0 until Board.COLUMNS) {
            symbol = checkColumns(blocks, col)
            symbol?.let { setWinner(it) }
        }
        symbol = checkDiagonals(blocks)
        symbol?.let { setWinner(it) }
    }

    private fun checkRow(blocks: Array<Array<String?>>, rowPos: Int): String? {
        val row = blocks[rowPos]
        val symbol = row.first()
        val isLine = row.all { it == symbol }
        return if (isLine) symbol else null
    }

    private fun checkColumns(blocks: Array<Array<String?>>, columnPos: Int): String? {
        val symbol = blocks[0][columnPos]
        val isLine = (1 until blocks.size).all { pos -> blocks[pos][columnPos] == symbol }
        return if (isLine) symbol else null
    }

    private fun checkDiagonals(blocks: Array<Array<String?>>): String? {
        // First diagonal
        var symbol = blocks[0][0]
        var isLine = (1 until blocks.size).all { pos -> blocks[pos][pos] == symbol }
        if (isLine) return symbol
        // Second diagonal
        symbol = blocks[0][blocks.size - 1]
        isLine = (1 until blocks.size).all { pos -> blocks[pos][blocks.size - 1 - pos] == symbol}
        return if (isLine) symbol else null
    }

    private fun setWinner(symbol: String) = when (symbol) {
        player1Symbol -> board.status = BoardStatus.PLAYER_1_WINS
        player2Symbol -> board.status = BoardStatus.PLAYER_2_WINS
        else -> Unit
    }
}
